### PR TITLE
perf(statusline): UI elements are always redrawn on K_EVENT

### LINKED ISSUE #4847: statusline is evaluated too often (test case with deoplete)
I've noticed that `statusline` gets evaluated far too often, in combination with deoplete's init procedure.

- Neovim version: v0.1.5-329-g748898b
- [ ] Vim behaves differently? Vim version:

### Steps to reproduce using `nvim -u NORC`

vimrc.minimal:
```
set rtp+=~/.vim/neobundles/deoplete
let g:deoplete#enable_at_startup = 1

let s:c = 0
fun! StatusLine()
	let s:c += 1
	return s:c
endfun
set statusline=%!StatusLine()

" To trigger deoplete's init.
set updatetime=200
```

```
nvim -u vimrc.minimal vimrc.minimal
```

After loading the file, the counter will be at ~2.
Now reload the file using `:e`: the counter will be at ~280 - 330!

From poking in Neovim's source it seems to happen through `normal_redraw` in `normal_check`.

It seems to depend on the duration of the statusline function: adding `sleep 500m` there does not cause it.

Unfortunately deoplete's init procedure is not really small, but that's where I've noticed it when trying to figure out where all the calls in airline are coming from.

### LINKED ISSUE #14303: Statusline gets updated on every call from unrelated timer
When there is a timer executing some function (unrelated to statusline), statusline gets updated on every call to timer function.

- `nvim --version`: NVIM v0.5.0-dev+1221-g9b2d4ff62
- `vim --version`: 8.0.1453
- Operating system/version: Ubuntu 18.04.2
- Terminal name/version: Gnome, 3.28.2-1ubuntu1~18.04.1
- `$TERM`: xterm-256color

### Steps to reproduce

I have written two config files: one uses Vimscript timer (`timer_start`) and the other uses Lua timer (`vim.loop.new_timer`). Both of them have statusline set to executable function which returns the number of times it has been called. In this setup number in statusline shows the number of times `statusline` option got evaluated since the start.

Steps:

- Run `nvim -u init_timer_vim.vim` and look at statusline.
- Run `nvim -u init_timer_lua.vim` and look at statusline.
- Run `vim -u init_timer_vim.vim` and look at statusline.

<details><summary>init_timer_vim.vim</summary>

```vim
" Set up global counter
let g:statusline_counter = 0

" Create statusline function
function MyStatusline()
  let g:statusline_counter = g:statusline_counter + 1
  return g:statusline_counter
endfunction

" Enable settings
set laststatus=2
set statusline=%!MyStatusline()

" Create timer which is unrelated to statusline settings
function MyHandler(timer)
endfunction

let timer = timer_start(200, 'MyHandler', {'repeat': 500})

" Results:
" - Neovim: just after opening, `g:statusline_counter` gets updated forever
"   every 200 ms, even though timer has nothing to do with statusline.
" - Vim: statusline `g:statusline_counter` isn't constantly updating.
```

</details>

<details><summary>init_timer_lua.vim</summary>

```vim
" Set up global counter
let g:statusline_counter = 0

" Create statusline function
function MyStatusline()
  let g:statusline_counter = g:statusline_counter + 1
  return g:statusline_counter
endfunction

" Enable settings
set laststatus=2
set statusline=%!MyStatusline()

" Create timer which is unrelated to statusline settings
lua << EOF
local timer = vim.loop.new_timer()
timer:start(200, 200, vim.schedule_wrap(function() end))
EOF

" Result: just after opening Neovim `g:statusline_counter` gets updated
" forever every 200 ms, even though timer has nothing to do with statusline.
```

</details>

### Actual behaviour

- Output of `nvim -u init_timer_vim.vim`:
<details><summary>Animation of Neovim with Vimscript timer</summary>

neovim_statusline-issue_1

</details>

- Output of `nvim -u init_timer_lua.vim`:
<details><summary>Animation of Neovim with Lua timer</summary>

neovim_statusline-issue_2

NOTE:  in reality updates were smooth, every 200ms. Certain "lags" are because of recording.

</details>

- Output of `vim -u init_timer_vim.vim`:
<details><summary>Animation of Vim with Vimscript timer</summary>

neovim_statusline-issue_3

</details>

### Expected behaviour

Maybe this is by design, but my expectation is that statusline shouldn't update on every call to unrelated timer (like in Vim).

### LINKED ISSUE #17937: High CPU usage when vim.schedule inside statusline
### Neovim version (nvim -v)

NVIM v0.6.1

### Vim (not Nvim) behaves the same?

vim doesn't support lua

### Operating system/version

Arch linux

### Terminal name/version

alacritty 0.10.0-dev (f90dd12e)

### $TERM environment variable

xterm-256color

### Installation

System package manager

### How to reproduce the issue

Create file `example.lua` with config:
```lua
local line = nil
function _G.make_statusline()
    vim.schedule(function()
        line = 'some status line'
    end)
    return line
end
vim.cmd([[autocmd BufWinEnter,WinEnter * setlocal statusline=%!v:lua.make_statusline()]])
```
Run nvim with the config:
```bash
$ nvim -u example.lua
```

### Expected behavior

Almost zero CPU usage.

### Actual behavior

Got 100% of one CPU core usage.

### PR DESCRIPTION
**Pull request by luukvbaal**
Problem:    'statusline'-format UI elements are redrawn on each K_EVENT.
Solution:   Only redraw UI elements when something relevant has changed.

Fix https://github.com/neovim/neovim/issues/4847 (untested)
Fix https://github.com/neovim/neovim/issues/14303
Fix https://github.com/neovim/neovim/issues/17937

### CODE REVIEW

**Comment by luukvbaal** [comment]
Set a global variable to avoid redrawing the `'statusline'`-format UI elements. I won't pretend to understand all the implications of doing so (unfamiliar with the event loop). All tests still pass however after unsetting the boolean for cases where elements should explicitly be redrawn. Not sure if any untested cases are affected.

File: src/nvim/ex_docmd.c [hunk 1]
```diff
@@ -6156,6 +6156,7 @@ static void ex_redrawstatus(exarg_T *eap)
     status_redraw_curbuf();
   }
 
+  skip_redr_custom_k_event = false;
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:6159]: This is also needed in `ex_redrawtabline()`

**Comment by zeertzjq** [comment]
With this PR, sourcing the following Lua file causes the tabline to disappear:
```lua
vim.g.var = 'foo'
vim.o.showtabline = 2
vim.o.tabline = '%!g:var'
vim.schedule(function()
  vim.g.var = 'bar'
  vim.api.nvim_input('<C-L>')
end)
```

**Comment by luukvbaal** [comment]
> With this PR, sourcing the following Lua file causes the tabline to disappear:

Fixed (by adding by unsetting the variable in the `<C-L>` code path). Does this seem like the correct approach if we can catch all the cases where we should ignore/unset `skip_redr_custom_k_event`?

**Comment by luukvbaal** [comment]
> > With this PR, sourcing the following Lua file causes the tabline to disappear:

> Fixed (by unsetting the variable in the <C-L> code path).

The tabline is still cleared however (and not redrawn) for example when resizing the terminal. It seems to be cleared in `drawscreen.c:screenclear()` by `ui_call_grid_clear()`. 

I guess the tabline is part of the default grid and thus is cleared. Does it need to be cleared? Does it need to be part of the grid?

**Comment by luukvbaal** [comment]
Previous comment might be irrelevant as it is fixed by the latest commit🤷🏻

**Comment by justinmk** [comment]
Will be very nice to fix this, though adding another special-case global is unfortunate :)

File: src/nvim/statusline.c [hunk 2]
```diff
@@ -279,7 +279,8 @@ static void win_redr_custom(win_T *wp, bool draw_winbar, bool draw_ruler)
   // There is a tiny chance that this gets called recursively: When
   // redrawing a status line triggers redrawing the ruler or tabline.
   // Avoid trouble by not allowing recursion.
-  if (entered) {
+  if (entered || skip_redr_custom_k_event) {
+    skip_redr_custom_k_event = false;
```
Comment (justinmk) [inline src/nvim/statusline.c:283]: Related globals `redraw_tabline` and `redraw_cmdline` look very similar in nature. Would it make sense to name this one `redraw_win_custom` ?

**Comment by bfredl** [comment]
I don't think this is the right solution. There is nothing wrong with redrawing as a result of K_EVENT and this will_ break_ other plugins which use K_EVENT to respond directly to user intention (or respond external events in a proper fashion, not hot timer loop panic). 

Also we already have a set of variables to control _precisely_ what is going to be redrawn or not in the next redraw "tick". In particular, for a window we have `wp->w_redr_status` to control if the statusline is supposed to be redrawn. What we want is to avoid setting it to `true` in the first place. We need to find the precise code location that wants to set it to true in those linked issues and see if the condition can be made sharper so there is no redraw if no actual state has changed.

**Comment by zeertzjq** [comment]
`show_cursor_info()` redraws custom statusline even if `w_redr_status` isn't set.

**Comment by bfredl** [comment]
It is supposed only redraw the ruler or statusline when cursor was actually moved.

>  if false, only show ruler if position has changed.

But this only is actually checked for the builtin/default ruler or statusline. This check needs be moved and also apply when w_redr_custom redraws the ruler.

**Comment by luukvbaal** [comment]
> This check needs be moved earlier in show_cursor_info()

Yeah that does seem like a better fix 👍🏻 Let me see if I can make that refactor without breaking any tests :)

**Comment by luukvbaal** [comment]
> also, ideally, even this should be done as a check before update_screen() in normal_redraw() and ins_redraw() which sets the shared flag and then use the normal redraw entry point.

I briefly tried this but my attempt introduced too many test failures.

> This check needs be moved earlier in show_cursor_info()

Latest commit implements this. There are two remaining test failures that I can't really explain.

For some reason this change seems to break the `wordcount()` function🤡
And the "statusline is redrawn on entering cmdline with scrolled messages" on `test/functional/ui/cmdline_spec.lua:891` fails for some reason. The one that closes the scrolled message, the tests before it work fine. Don't really see how that matters for the code that handles this in `ex_getln.c:789`.

**Comment by bfredl** [comment]
the failure in ` test/functional/ui/cmdline_spec.lua` is real. as we redraw the statusline when entering cmdline mode, thus redrawing it when exiting cmdline is reasonably to be expected. this was done implicitly before, but we need to do this explicitly now, maybe by setting `curwin->w_redr_status` in `ex_colon` but only if `getexline` was used.

**Comment by luukvbaal** [comment]
> maybe by setting curwin->w_redr_status in ex_colon but only if getexline was used.

Adding a `status_redraw_all()` at the end of `command_line_enter()` fixed this (did that used to be called `ex_colon()`?). I don't think it needs to be done conditionally for `curwin` only? It is done unconditionally for all windows at the start of `command_line_enter()` as well.

Adding a `w_ru_state` that stores `get_real_state()` and checking for that in `show_cursor_info()` fixed the `wordcount()` test failure (those tests use `'statusline'` to store the wordcount in a variable so I guess the failing test was not that 🤡 after all).

I guess all these explicit statusline redraws that have to be added now reveals that `show_cursor_info()` redrawing unconditionally was responsible for updating the status line where it really shouldn't have been? I guess there might be other cases affected as well that have not been caught by the test-suite.

**Comment by bfredl** [comment]
> I guess all these explicit statusline redraws that have to be added now reveals that show_cursor_info() unconditionally redrawing was responsible for updating the status line where it really shouldn't have been? I suppose there might be other cases affected as well that have not been caught by the test-suite.

Yes, this will likely affect some async plugins. But this will be more of an annoyance than a critical bugs. As issues get reported to us for these cases, we can then consider them individually which should be fixed in neovim, or in the plugin (by redrawing status explicit)

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -908,6 +908,7 @@ static uint8_t *command_line_enter(int firstc, long count, int indent, bool clea
     msg_ext_clear_later();
   }
 
+  status_redraw_all();  // redraw to show mode change
```
[the remark this answers is not in the document]
Comment (luukvbaal) [reply src/nvim/ex_getln.c:911]: That was my initial instinct as well but it didn't resolve the test failure(`test/functional/ui/cmdline_spec.lua:891`).

[the remark this answers is not in the document]
Comment (luukvbaal) [reply src/nvim/ex_getln.c]: Sorry, I lied. Seem to have remembered this wrong, it passes now.

**Comment by luukvbaal** [comment]
> Yes, this will likely affect some async plugins. But this will be more of an annoyance than a critical bugs. As issues get reported to us for these cases, we can then consider them individually which should be fixed in neovim, or in the plugin (by redrawing status explicit)

Do you think we can still include this for 0.9 then? Perhaps accompanied with a `news.txt` item pointing out the change/workaround?

**Comment by bfredl** [comment]
yes. The bug being fixed here is much more critical, so it will go in 0.9.

**Comment by clason** [comment]
Also, any followup fixes can be backported; we usually put out a 0.x.1 bugfix release fairly quickly.

File: src/nvim/drawscreen.c [hunk 4]
```diff
@@ -733,28 +734,53 @@ void show_cursor_info(bool always)
...
+    if (*p_wbr != NUL || *curwin->w_p_wbr != NUL) {
+      win_redr_winbar(curwin);
+    }
 
-  if (need_maketitle
-      || (p_icon && (stl_syntax & STL_IN_ICON))
-      || (p_title && (stl_syntax & STL_IN_TITLE))) {
-    maketitle();
-  }
+    if (need_maketitle
```
Comment (luukvbaal) [inline src/nvim/drawscreen.c:764]: Why are this and the `draw_tabline()` below conditional on `need_maketitle/redraw_tabline`? Either we want to support showing the cursor(now mode) information in the title/tabline or we don't. If we do they should be unconditionally called here I think, `set tal=%{mode()}` will not show the mode changes for example. Them being present in `show_cursor_info()` seems to indicate that we want to support this.

Comment (bfredl) [reply src/nvim/drawscreen.c:764]: I agree it is weird, but let's consider that in a follow up PR (as I said, I want to refactor a lot of this code anyway)

Comment (bfredl) [reply src/nvim/drawscreen.c:764]: although maybe the alternate `stl_syntax & STL_IN_TITLE)` condition handles this? too many magic variables..

Comment (luukvbaal) [reply src/nvim/drawscreen.c:764]: Yeah for the title maybe. But like I said set `tal=%{mode()}` will not show mode changes due to the guard on `redraw_tabline` here.

Another weird thing: there is a `maketitle()` in `normal_redraw()` right before `show_cursor_info()` which will be redundant depending on what we do with this.

**Pull request merged**